tests/net/proptest.rs

```diff
@@ -0,0 +1,177 @@
+//! Proptest helpers and strategies
+//!
+//! This module houses strategies to generate (and reduce/expand) various `hbbft` and `net` related
+//! structures.
+
+use proptest::prelude::Rng;
+use proptest::strategy::{Strategy, ValueTree};
+use proptest::test_runner::{Reason, TestRunner};
+
+/// Node network topology.
+///
+/// A `NetworkTopology` describes the number of correct and faulty nodes in a network. It can also
+/// be checked, "averaged" (using the `halfway` function) and generated using `NetworkTopologyTree`.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NetworkTopology {
+    /// Total number of nodes in network.
+    pub size: usize,
+    /// Number of faulty nodes in a network.
+    pub faulty: usize,
+}
+
+impl NetworkTopology {
+    /// Creates a new `NetworkTopology` with the supplied parameters.
+    ///
+    /// Topologies that are not sane can be created using this function.
+    pub fn new(size: usize, faulty: usize) -> Self {
+        NetworkTopology { size, faulty }
+    }
+
+    /// Checks whether the network topology satisfies the `3 * faulty + 1 <= size`
+    pub fn is_sane(&self) -> bool {
+        self.faulty * 3 + 1 <= self.size
+    }
+
+    /// Creates a new topology of average complexity.
+    ///
+    /// The new topology is approximately half way in the interval of `[self, high]` and will
+    /// conform to the constraint checked by `is_sane()`.
+    ///
+    /// # Panics
+    ///
+    /// `high` must be have a higher or equal size and faulty node count.
+    pub fn halfway(&self, high: NetworkTopology) -> NetworkTopology {
```

Comment:
As you mention in "Panics", the function doesn't work with an `other` that has a lower faulty node count. Maybe rename `halfway` to something like `average_higher` to reflect that?